ledger/participant-integration-api/src/main/scala/platform/store/state/StateCache.scala

```diff
@@ -0,0 +1,122 @@
+package com.daml.platform.store.state
+
+import java.util.concurrent.atomic.{AtomicLong, AtomicReference}
+
+import com.daml.caching.Cache
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.store.state.StateCache.PendingUpdates
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+trait StateCache[K, V] {
+  protected implicit def ec: ExecutionContext
+  protected def cache: Cache[K, V]
+
+  protected val logger: ContextualizedLogger = ContextualizedLogger.get(getClass)
+
+  private[store] val pendingUpdates =
+    collection.concurrent.TrieMap.empty[K, PendingUpdates]
+
+  def get(key: K)(implicit loggingContext: LoggingContext): Option[V] =
+    cache.getIfPresent(key) match {
+      case Some(value) =>
+        logger.debug(s"Cache hit for $key -> ${value.toString.take(100)}")
+        Some(value)
+      case None =>
+        logger.debug(s"Cache miss for $key ")
+        None
+    }
+
+  def put(key: K, validAt: Long, value: V)(implicit
+      loggingContext: LoggingContext
+  ): Unit = putAsync(key, validAt, Future.successful(value))
+
+  def putAsync(key: K, validAt: Long, eventualValue: Future[V])(implicit
+      loggingContext: LoggingContext
+  ): Unit = {
+    logger.debug(s"New pending cache update for key $key at $validAt")
+    val pendingUpdate @ PendingUpdates(pendingCountRef, latestRef, effectsChain) =
+      pendingUpdates.getOrElseUpdate(key, PendingUpdates.empty)
```

Comment:
Reminder from our discussion: add a note on why we need the lock-free version of this cache, and where the benchmark can be found that measures its maximum throughput when accessing different keys concurrently.